nxcomp/Log.h

```diff
@@ -92,27 +96,61 @@ class NXLogStamp
 class NXLog
 {
 	NXLogLevel level_;
-	NXLogLevel current_level_;
-	std::string current_file_;
 
 	std::ostream *stream_;
 	std::map< std::string, NXLogLevel > per_file_levels_;
+	bool synchronized_;
+	int  thread_buffer_size_;
+	pthread_mutex_t output_lock_;
+
+	pthread_key_t tls_key_;
+
+	typedef struct per_thread_data_s
+	{
+		NXLogLevel          current_level;
+		std::string*        current_file;
+		std::string*        thread_name;
+		std::stringstream*  buffer;
+	} per_thread_data;
+
+
+	per_thread_data* get_data()
+	{
+		per_thread_data *ret = NULL;
+ 
+		if ( (ret = (per_thread_data*)pthread_getspecific(tls_key_)) == NULL )
+		{
+			ret = (per_thread_data*)malloc(sizeof(*ret));
+			ret->current_level = NXDEBUG;
+			ret->buffer = new std::stringstream();
+			ret->current_file = new std::string();
+			ret->thread_name = new std::string();
+			pthread_setspecific(tls_key_, ret);
+		}
+
+		return ret;
+	}
 
 	public:
 	NXLog()
 	{
-		stream_        = &std::cerr;
-		level_         = NXDEBUG;
-		current_level_ = NXDEBUG;
+		stream_             = &std::cerr;
+		level_              = NXDEBUG;
+		synchronized_       = true;
+		thread_buffer_size_ = 1024;
+		pthread_key_create(&tls_key_, free);
+
 	}
 
 	NXLog(char *filename);
 
 	NXLog(std::ostream &stream)
 	{
-		stream_        = &stream;
-		level_         = NXDEBUG;
-		current_level_ = NXDEBUG;
+		stream_             = &stream;
+		level_              = NXDEBUG;
+		synchronized_       = true;
+		thread_buffer_size_ = 1024;
+		pthread_key_create(&tls_key_, free);
```

Comment:
Error handling?

Btw, duplicated code, better make the parameter here optional and default to `std::cerr`, so we can get rid of NXLog().